testsuite/tests/uart.rs

```diff
@@ -132,43 +173,115 @@ mod tests {
+    #[test]
+    fn check_parity_mismatch(state: &mut super::State) {
+        // Repurpose crossover serials from state.
+        let (usart_even, pins_even) = unwrap!(state.serial_slow.take()).free();
+        let config_even = Config::default().parity(Parity::Even);
+        let mut serial_even = Serial::new(
+            usart_even,
+            pins_even,
+            config_even,
+            state.clocks,
+            &mut state.apb1
+        );
+        let (usart_odd, pins_odd) = unwrap!(state.serial_fast.take()).free();
+        let config_odd = Config::default().parity(Parity::Odd);
+        let mut serial_odd = Serial::new(
+            usart_odd,
+            pins_odd,
+            config_odd,
+            state.clocks,
+            &mut state.apb1
+        );
+
+        // Transmit data with wrong parity in both directions.
+        unwrap!(nb::block!(serial_even.write(b'x')));
+        let result = nb::block!(serial_odd.read());
+        assert!(matches!(result, Err(Error::Parity)));
+
+        unwrap!(nb::block!(serial_odd.write(b'x')));
+        let result = nb::block!(serial_even.read());
+        assert!(matches!(result, Err(Error::Parity)));
+
+        // Finally restore serial devices in state.
+        let (usart_slow, pins_slow) = serial_even.free();
+        let serial_slow = Serial::new(
+            usart_slow,
+            pins_slow,
+            BAUD_SLOW,
+            state.clocks,
+            &mut state.apb1
+        );
+        let (usart_fast, pins_fast) = serial_odd.free();
+        let serial_fast = Serial::new(
+            usart_fast,
+            pins_fast,
+            BAUD_FAST,
+            state.clocks,
+            &mut state.apb1
+        );
+        state.serial_slow = Some(serial_slow);
+        state.serial_fast = Some(serial_fast);
+    }
+
     #[test]
     fn send_receive_wrong_baud(state: &mut super::State) {
         let (mut tx_slow, mut rx_slow) = unwrap!(state.serial_slow.take()).split();
         let (mut tx_fast, mut rx_fast) = unwrap!(state.serial_fast.take()).split();
 
         // provoke an error (framing)
-        defmt::unwrap!(nb::block!(tx_slow.write(b'a')));
+        unwrap!(nb::block!(tx_slow.write(b'a')));
         let c = nb::block!(rx_fast.read());
-        defmt::info!("{}", c);
+        defmt::debug!("{}", c);
         assert!(matches!(c, Err(Error::Framing)));
 
         // provoke an error (this does not seem to be absolutely deterministic
-        // and we've seen multiple error variants in the wild)
+        // and we've seen multiple error variants in the wild, including
+        // receiving the wrong but valid character)
         unwrap!(nb::block!(tx_fast.write(b'a')));
-        let c = nb::block!(rx_slow.read());
-        defmt::info!("{}", c);
-        assert!(matches!(c, Err(Error::Framing) | Err(Error::Noise)));
+        let result = nb::block!(rx_slow.read());
+        defmt::debug!("{}", result);
+        assert!(match result {
+            Ok(c) => c != b'a',
```

Comment:
I'd still would like a warning here. Because when executing the test there is no indication about how the test succeeded. 

```suggestion
            Ok(c) => {
                defmt::warn!("Expected Err, got Ok({:x})", c);
                c != b'a'
            },
```